sys/usb/usbus/usbus.c

```diff
@@ -0,0 +1,360 @@
+        ep->ep = usbdev_ep;
+        if (iface) {
+            ep->next = iface->ep;
+            iface->ep = ep;
+        }
+    }
+    return ep;
+}
+
+static inline uint32_t _get_ep_bitflag(usbdev_ep_t *ep)
+{
+    /* Bit flag, lower half are OUT endpoints, upper half are IN endpoints */
+    return 1 << ((ep->dir == USB_EP_DIR_IN ? 0x10 : 0x00) + ep->num);
+}
+
+static uint32_t _get_and_reset_ep_events(usbus_t *usbus)
+{
+    unsigned state = irq_disable();
+    uint32_t res = usbus->ep_events;
+
+    usbus->ep_events = 0;
+    irq_restore(state);
+    return res;
+}
+
+static void _signal_handlers(usbus_t *usbus, uint16_t flag,
+                             uint16_t msg)
+{
+    for (usbus_handler_t *handler = usbus->handlers;
+         handler; handler = handler->next) {
+        if (handler->flags & flag) {
+            handler->driver->event_handler(usbus, handler, msg);
+        }
+    }
+}
+
+static void _usbus_init_handlers(usbus_t *usbus)
+{
+    for (usbus_handler_t *handler = usbus->handlers;
+         handler; handler = handler->next) {
+        handler->driver->init(usbus, handler);
+    }
+}
+
+static void *_usbus_thread(void *args)
+{
+    usbus_t *usbus = (usbus_t *)args;
+    usbus_control_handler_t ep0_handler;
+
+    event_queue_init(&usbus->queue);
+
+    usbus->control = &ep0_handler.handler;
+
+    usbus_control_init(usbus, &ep0_handler);
+
+    usbdev_t *dev = usbus->dev;
+    usbus->pid = sched_active_pid;
+    usbus->addr = 0;
+    usbus->iface = NULL;
+    usbus->str_idx = 1;
+    DEBUG("usbus: starting thread %i\n", sched_active_pid);
+    /* setup the link-layer's message queue */
+    /* register the event callback with the device driver */
+    dev->cb = _event_cb;
+    dev->epcb = _event_ep_cb;
+    /* initialize low-level driver */
+    dev->context = usbus;
+    usbdev_init(dev);
+
+    usbus_add_string_descriptor(usbus, &usbus->config,
+                                USB_CONFIG_CONFIGURATION_STR);
+    usbus_add_string_descriptor(usbus, &usbus->product, USB_CONFIG_PRODUCT_STR);
+    usbus_add_string_descriptor(usbus, &usbus->manuf, USB_CONFIG_MANUF_STR);
+
+    usbus->state = USBUS_STATE_DISCONNECT;
+
+    /* Initialize handlers */
+    _usbus_init_handlers(usbus);
+
+#if (USBUS_AUTO_ATTACH)
+    static const usbopt_enable_t _enable = USBOPT_ENABLE;
+    usbdev_set(dev, USBOPT_ATTACH, &_enable,
+               sizeof(usbopt_enable_t));
+#endif
+
+    while (1) {
+        thread_flags_t flags = thread_flags_wait_any(
+            USBUS_THREAD_FLAG_USBDEV |
+            USBUS_THREAD_FLAG_USBDEV_EP |
+            THREAD_FLAG_EVENT
+            );
+        if (flags & USBUS_THREAD_FLAG_USBDEV) {
+            usbdev_esr(dev);
+        }
+        if (flags & USBUS_THREAD_FLAG_USBDEV_EP) {
+            uint32_t events = _get_and_reset_ep_events(usbus);
+            while (events) {
+                unsigned num = bitarithm_lsb(events);
+                events &= ~(1 << num);
+                if (num >= 0x10) {
```

Comment:
You should explain the value.

Reply:
Reworked a bit, should be explained now.